File: frontend/src/lib/readable-id.ts
/**
 * Shared readable ID generation using CSPRNG (crypto.getRandomValues).
 */

const CHARS = "abcdefghijklmnopqrstuvwxyz0123456789";

/**
 * Generate a cryptographically random alphanumeric suffix.
 *
 * @param length - Length of the suffix (default 6)
 * @returns Random lowercase alphanumeric string
 */
export function generateRandomSuffix(length = 6): string {
  const limit = CHARS.length;
  // Largest multiple of `limit` that fits in a Uint32.  Values at or
  // above this threshold would produce modulo bias and are discarded.
  const maxUnbiased = limit * Math.floor(0x100000000 / limit);
  const buf = new Uint32Array(length);
  let result = "";
  let filled = 0;
  while (filled < length) {
    crypto.getRandomValues(buf);
    for (let j = 0; j < buf.length && filled < length; j++) {
      if (buf[j] < maxUnbiased) {
        result += CHARS[buf[j] % limit];
        filled++;
      }
    }
  }
  return result;
}

/**
 * Sanitize a name into a readable ID base (without suffix).
 *
 * Transforms to lowercase, replaces spaces with hyphens, strips
 * special characters.
 *
 * @param name - Human-readable name to transform
 * @returns Sanitized slug, or empty string if name is blank
 */
export function generateReadableIdBase(name: string): string {
  if (!name || name.trim() === "") return "";

  let readable_id = name.toLowerCase().trim();

  // Remove characters that aren't letters, numbers, or spaces
  readable_id = readable_id.replace(/[^a-z0-9\s]/g, "");

  // Replace spaces with hyphens
  readable_id = readable_id.replace(/\s+/g, "-");

  // Collapse consecutive hyphens
  readable_id = readable_id.replace(/-+/g, "-");

  // Trim leading/trailing hyphens
  readable_id = readable_id.replace(/^-|-$/g, "");

  return readable_id;
}

/**
 * Generate a complete readable ID from a name.
 *
 * Combines the sanitized base with a random suffix, e.g.
 * `"finance-data-ab12x9"`.
 *
 * @param name - Human-readable name
 * @returns Full readable ID, or empty string if name is blank
 */
export function generateReadableId(name: string): string {
  const base = generateReadableIdBase(name);
  if (!base) return "";
  return `${base}-${generateRandomSuffix()}`;
}

File: backend/airweave/core/readable_id.py
"""Readable ID generation using a CSPRNG.

Provides a single ``generate_readable_id`` helper that converts a
human-readable name into a URL-safe slug with a random suffix, e.g.
``"finance-data-ab12x9"``.  The suffix is drawn from
:func:`secrets.choice` (CSPRNG) rather than :mod:`random` (Mersenne
Twister).
"""

import re
import secrets
import string

_ALPHABET = string.ascii_lowercase + string.digits


def generate_readable_id(name: str) -> str:
    """Generate a readable ID from a name.

    Converts the name to lowercase, replaces spaces with hyphens,
    removes special characters, and appends a cryptographically random
    6-character suffix to ensure uniqueness.

    Args:
        name: The human-readable name to convert.

    Returns:
        A URL-safe readable identifier (e.g. ``"finance-data-ab123x"``).
    """
    # Convert to lowercase and replace spaces with hyphens
    readable_id = name.lower().strip()

    # Replace any character that's not a letter, number, or space with nothing
    readable_id = re.sub(r"[^a-z0-9\s]", "", readable_id)
    # Replace spaces with hyphens
    readable_id = re.sub(r"\s+", "-", readable_id)
    # Ensure no consecutive hyphens
    readable_id = re.sub(r"-+", "-", readable_id)
    # Trim hyphens from start and end
    readable_id = readable_id.strip("-")

    # Add random alphanumeric suffix (CSPRNG)
    suffix = "".join(secrets.choice(_ALPHABET) for _ in range(6))
    readable_id = f"{(readable_id + '-') if readable_id else ''}{suffix}"

    return readable_id
